Name: Oliver Vesey
<br>
Student ID: 101043434
<br>
$\beta = 7$
<br>
$C = 12$




In [4]:
import numpy as np
import matplotlib.pyplot as plt
%matplotlib inline
from scipy.stats import t
from scipy import stats 

class SLR: 
    
 
    def __init__(self, x_vals, y_vals):   #This initalises our SLR Class and stores the x-values and y-values in seperate NumPy arrays.
        self.x_vals = np.array(x_vals, dtype = float)
        self.y_vals = np.array(y_vals, dtype = float)
        self.lx = len(self.x_vals)  #lx is the length of x_vals (so also y_vals), and is useful in methods below.
        self.x_bar = np.mean(self.x_vals) #Calculates our sample means, x-bar and y-bar to be used later.
        self.y_bar = np.mean(self.y_vals)
        if len(self.x_vals) != len(self.y_vals):
            raise ValueError("There must be an equal amount of x and y data values.") #Checks to ensure there are an equal amount of x and y values.
        
#Method 1i)
        
        
    def AB(self):
        Sxx = 0.0
        Sxy = 0.0
        for i in range(self.lx):
            Sxx += ((self.x_vals[i]-self.x_bar)**2)
            Sxy += ((self.x_vals[i]-self.x_bar)*(self.y_vals[i]-self.y_bar)) #Calculates Sxx and Sxy using the formulas in a for loop.
        B = float(Sxy/Sxx)
        A = float(self.y_bar - B*(self.x_bar))
        return(A,B)  #Calculates and returns a list of two items A, then B.
    

#Method 1ii)

    def scatterxy(self):
        plt.scatter(self.x_vals, self.y_vals,label = "Data",color = "blue")
        plt.xlabel("$x_i$")
        plt.ylabel("$y_i$")
        plt.title("Scatter plot of y values against x values")
        plt.legend()
        plt.show()  

        
#Method 1iii)

    def SRL(self, x): 
        a = self.AB()[0] #Stores a and b as the first and second items of the returned value of AB using the original x and y values.
        b = self.AB()[1]
        value = a + b*x 
        return(value) #Returns the value of the sample regression line at the point x.
    
    
#Method 1iv)

    def MSE(self):
        SSE = 0.0
        for a in range(self.lx):
            y_i = self.SRL(self.x_vals[a])  #Finds the fitted value yi.
            SSE += (self.y_vals[a] - y_i)**2  #Sums to find the value of SSE.
        return(SSE/(self.lx - 2))
               
               
#Method 1v)
               
    def R2(self):
        SST = 0.0
        SSE = 0.0
        for a in range(self.lx):
            y_i = self.SRL(self.x_vals[a])  #Finds the fitted value yi.
            SSE += (self.y_vals[a] - y_i)**2  #Sums to find the value of SSE and SST respectively.
            SST += (self.y_vals[a] - self.y_bar)**2
        return((SST-SSE)/SST)
               
#Method 1vi)  
            
            
    def PVAL(self):
        B = self.AB()[1]    #Retrieves B value from AB method.
        s2 = self.MSE()     #s^2 is the MSE value.
        Sxx = 0.0
        for i in range(self.lx):      #Computes Sxx directly as efore.
            Sxx += ((self.x_vals[i]-self.x_bar)**2)      
        Tobs = B/(np.sqrt(s2/Sxx))       #Observed t-statistic for H0: B = 0.                                
        p_val = 2*(1 - t.cdf(abs(Tobs), self.lx - 2))    #Two-sided p-value using the permitted SciPy t-distribution.
        return(p_val)    #Returns the p-value.
        




In [6]:
np.random.seed(101043434)

n=30
C=12
alpha=2025.0
beta=7
sigmas = [3,17,28] #'small', 'moderate' and 'large' sigma values.


x=C*np.random.uniform(1,2,n) #Create x-values

for i in sigmas:
    err = i*np.random.normal(0,1,n)  #Creating the different error and corresponding y alues for each sigma value.
    y=-beta*x+alpha+err
    model = SLR(x,y)  #First finding the different values with my code.
    print("Simulation of the data with sigma =",i,":")
    print("Alpha from my code is ",model.AB()[0])
    print("Beta from my code is ",model.AB()[1])
    print("The p-value from my code is ",model.PVAL())
    print("The R^2 value from my code is ",model.R2())
    print("")
    
    res = stats.linregress(x, y) #Here finding the different values with stats.linregress.
    print("Alpha from stats.linregress is ",res.intercept)
    print("Beta from stats.linregress is ",res.slope)
    print("The p-value from stats.linregress is ",res.pvalue)
    print("The R^2 value from stats.linregress is ",(res.rvalue)**2)
    print("")
    print("")
    


Simulation of the data with sigma = 3 :
Alpha from my code is  2028.2035026744788
Beta from my code is  -7.139457162147309
The p-value from my code is  0.0
The R^2 value from my code is  0.9759779442436807

Alpha from stats.linregress is  2028.2035026744788
Beta from stats.linregress is  -7.139457162147311
The p-value from stats.linregress is  3.2207052887904475e-24
The R^2 value from stats.linregress is  0.9759779442436809


Simulation of the data with sigma = 17 :
Alpha from my code is  2024.8167565679992
Beta from my code is  -7.105611130939468
The p-value from my code is  2.760627282327732e-10
The R^2 value from my code is  0.7643198022513203

Alpha from stats.linregress is  2024.8167565679992
Beta from stats.linregress is  -7.105611130939468
The p-value from stats.linregress is  2.760627542311948e-10
The R^2 value from stats.linregress is  0.7643198022513203


Simulation of the data with sigma = 28 :
Alpha from my code is  2019.7462586227205
Beta from my code is  -6.61541002727214

## Report and Summary of Code vs Built-in:
<br>
Above is a simulation of the data showing what my code produces, followed by what stats.linregress produces for $\sigma = 3$, $\sigma = 17$, and $\sigma = 28$.
<br>
<br>
For all three of my different small, moderate and large $\sigma$ values, $\alpha$, $\beta$, $R^2$, and the $p$-value calculated by my code and the in built in stats.linregress, are all extremely close to one another (certainly within at least 5dp) if not the same. For example, $\beta = -7.105611130939468$ when $\sigma$ = 17 for both my code and stats.linregress.
<br>
<br>
We can observe that $\alpha$, $\beta$ and $R^2$ decrease in size as $\sigma$ increases, starting with ($\sigma$ at smallest value), $$\alpha \approx 2028.20350 , \beta \approx -7.13945, R^2 \approx 0.97597$$ and finishing with ($\sigma$ at largest value), $$\alpha \approx 2019.74625, \beta \approx -6.61541, R^2 \approx 0.36901$$
<br>
Meanwhile, the $p$-value increases as $\sigma$ increases, starting with $p \approx 0$, and finishing with $p \approx 0.00037$. This data suggests as $\sigma$ increases we get a weaker linear regression.


## USER GUIDE

## Class Initialiser SLR

To use it, create an instance of the class by calling, for example, 'model = SLR(x, y)', with x and y being arrays of x values and y values respectively (e.g. x = [1,2,3], y = [4,5,6]. The number of x and y values must match otherwise an error will be raised. These will be saved as a NumPy array to make calculations easier later on.



## Method 1i – AB

This method calculates and returns the intercept $\hat{\alpha}$ (A), and slope $\hat{\beta}$ (B) of the simple linear regression line.
It loops through all data points to find these coefficients manually, and it returns a tuple (A, B).

To use it, once you've created an instance of the class (e.g. 'model' from above), you could call, for example, 'A, B = model.AB()'' which would save A and B as $\hat{\alpha}$ and $\hat{\beta}$ respectively. To see the values you would then need to print them. (e.g. 'print(A, B)')


<br>

## Method 1ii – scatterxy

This method plots the observed data points as a scatter plot of $y_i$ values against $x_i$ values. Continuing our example, simply call 'model.scatterxy()' to display the plot.


<br>

## Method 1iii – SRL

SRL(x) returns the predicted y-value at any chosen x, written in with the function. For example, calling 'Y = model.SRL(5)' would return the estimated $y$ when $x=5$. To see this $y$ value, again you would need to print: 'print(Y)'.


<br>

## Method 1iv – MSE

This method calculates and returns the Mean Squared Error (MSE). Continuing our example, you could call 'model.MSE()' to get the value of MSE, and print/save the value in the same fashion as previous methods. (e.g. 'print(model.MSE())')


<br>

## Method 1v – R2

This computes and returns the $R^2$ value. Call 'model.R2()' to get the $R^2$ value, and print/save the value in the same fashion as previous methods. (e.g. 'print(model.R2())')


<br>

## Method 1vi – PVAL

This method calculates and returns the $p$-value using the permitted t-distribution. As before, call 'model.PVAL()' to get the $p$-value, and print/save the value in the same fashion as previous methods. (e.g. 'print(model.PVAL())')





